src/main/java/com/sandvoxel/generitech/common/tileentities/machines/TileEntityPulverizer.java

```diff
@@ -201,153 +193,164 @@ public boolean canBeRotated() {
-                if (inventory.getStackInSlot(4).getItem() == lastFuelType) {
-                    fuelRemaining = lastFuelValue;
+    public boolean cancrush(ItemStack item) {
+        return PulverizerRegistry.containsInput(item);
+    }
 
+    public void burnTime(int i) {
+        if (fuelRemaining == 0) {
+            if (inventory.getStackInSlot(4).getItem() == lastFuelType && cancrush(inventory.getStackInSlot(i))) {
+                fuelRemaining = lastFuelValue;
 
-                } else {
-                    fuelRemaining = net.minecraft.tileentity.TileEntityFurnace.getItemBurnTime(inventory.getStackInSlot(4));
-                    lastFuelType = inventory.getStackInSlot(4).getItem();
-                    lastFuelValue = fuelRemaining;
 
-                }
-                fuelTotal = fuelRemaining;
-                inventory.decrStackSize(4, 1);
-                this.markDirty();
-                this.markForUpdate();
+            } else if (inventory.getStackInSlot(4).getItem() != lastFuelType && cancrush(inventory.getStackInSlot(i))) {
+                fuelRemaining = net.minecraft.tileentity.TileEntityFurnace.getItemBurnTime(inventory.getStackInSlot(4));
+                lastFuelType = inventory.getStackInSlot(4).getItem();
+                lastFuelValue = fuelRemaining;
             }
+            fuelTotal = fuelRemaining;
+            inventory.decrStackSize(4, 1);
+            this.markDirty();
+            this.markForUpdate();
         }
+    }
 
-        if ((this.container.takePower(powerUsage, true) == powerUsage && machineTier != MachineTier.TIER_0) || fuelRemaining > 0) {
-            if (machineActive && !pulverizerPaused) {
-                ticksRemaining--;
-
-                if (machineTier != MachineTier.TIER_0)
-                    this.container.takePower(powerUsage, false);
-                if (fuelRemaining > 0)
-                    fuelRemaining--;
-
-            }
-
-            if (inventory.getStackInSlot(0) != null && inventory.getStackInSlot(1) == null) {
-                ItemStack itemIn = inventory.getStackInSlot(0);
-                ItemStack itemOut;
-
-                if (!PulverizerRegistry.containsInput(itemIn))
-                    return;
+    @Override
+    public void update()
+    {
 
-                if (itemIn.stackSize - 1 <= 0) {
-                    itemOut = itemIn.copy();
-                    itemIn = null;
-                } else {
-                    itemOut = itemIn.copy();
 
-                    itemOut.stackSize = 1;
-                    itemIn.stackSize = itemIn.stackSize - 1;
-                }
+        //checking for the machine type
+        if (machineTier == null)
+            machineTier = MachineTier.byMeta(getBlockMetadata());
 
-                if (itemIn != null && itemIn.stackSize == 0) itemIn = null;
-                if (itemOut.stackSize == 0) itemOut = null;
+        if (!canWork()) {
+            machineActive = false;
+            this.markForUpdate();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "update" is 40 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.aadev.xyz/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)